### ISSUE TITLE
Support clustermq as backend for drake

### ISSUE THREAD

**Issue by wlandau** [mschubert/clustermq#86]
First of all, I am super excited about this package! I just tried it out, and it automatically detected my company's SGE scheduler and submitted jobs in a fraction of the time I am used to. Way cool!

I am thinking about a couple of `clustermq` backends for [`drake`](https://github.com/ropensci/drake): one for [persistent workers](https://ropenscilabs.github.io/drake-manual/hpc.html#persistent-scheduling), and another for [transient workers](https://ropenscilabs.github.io/drake-manual/hpc.html#transient-scheduling). The former is easiest to implement, but I do not expect to see performance gains there because persistent workers need to use the file system to save targets on the go.

However, `clustermq`-based transient workers could really shine, especially when compared to `drake`'s current [`future`/`batchtools`-based transient workers (`make(parallelism = "future")`](https://ropenscilabs.github.io/drake-manual/hpc.html#remote-workers), which add noticeable overhead to every single target. But to implement transient workers with `clustermq`, I believe I will need a way to submit and monitor workers in a non-blocking way. With `future`, for example, the master process can submit a job, go check on other jobs, and come back and check if the first job finished.

I suppose one potential alternative is to take advantage of the [native pipeline functionality in the schedulers themselves](https://hpc.nih.gov/docs/job_dependencies.html). Are you considering this for `clustermq`?

**Comment by wlandau** [comment]
Thanks for explaining, Michael. I think https://github.com/HenrikBengtsson/future/issues/204 would be ideal, and I am optimistic about its eventual implementation. I will close this issue as a duplicate of #23.

**Comment by mschubert** [comment]
I am reopening this for `drake` support specifically

**Comment by wlandau** [comment]
And I appreciate it. Really glad you're willing to help me speed up `drake`.

**Comment by mschubert** [comment]
Let me expand on the previous point a bit:

(1) I'm pretty sure you *do not want transient workers* for `drake` on the cluster (as in, a worker per call)

Many of your function calls may be short. Submitting many short-running jobs in a short period of time puts stress on the scheduler. The lower limit that is usually quoted is 2 minutes, and `snakemake` was scolded by sysadmins for using this exact approach (that they later tried to mend using job groups)

(2) There are usually batches of function calls where some arguments are iterated but others are not

Often, function calls will work on the same data but e.g. different indices are supplied. In this case, you will not want the common part to be sent over the network for each call separately.

I think the way to go here is to:

* Let `drake` define a pool of workers that will be common to the workflow (I think you already do this, but also allow for a different number of workers during those steps)
* Group tasks per function to call, organise iterated and constant arguments; if there are parallel chains, create a temporary function for the chain (I think you already do this or something similar)
* Let `clustermq` work on each of these groups one after another (this is where the parallel `foreach` or the `future` backend comes in)

`Q` in this case would be handled like `parLapply`, which is blocking itself but processes tasks in parallel.

The downside of course is that each group of function calls needs to be completed before the next can start. As the calls themselves are load-balanced, I don't think this is a big deal.

**Comment by wlandau** [comment]
You bring up a good point re transient workers. I do eventually want to get to job groups, but it is much harder to implement than persistent and transient workers. All in good time.

I do like the idea of maintaining a pool of common workers and submitting jobs to them as targets become ready (i.e. when their dependencies are all checked and built). The only downside is that users may hit wall time limits fairly easily, but we can work around that later with job groups (or simply by refreshing workers periodically).

Unfortunately, it is not so simple to group tasks by function call. Users define arbitrary dependency networks of commands, and those commands are arbitrary chunks of R code (not necessarily neat function calls). I suppose some targets sometimes share all the same dependencies, but I do not think such groupings would efficiently generalize to larger workflows.

> The downside of course is that each group of function calls needs to be completed before the next can start. 

At first, I did think this was such a big deal. What you describe is essentially what I have been calling "staged scheduling", and it was a good start. However, during its first several months of use, there were several projects that lost considerable parallel efficiency. `drake` could not always jump ahead in the graph even when targets were ready. https://github.com/ropensci/drake/issues/168 helped somewhat, but it was not nearly sufficient. I touched on the downsides of staged parallelism in a recent [rOpenSci tech note](https://ropensci.org/technotes/2018/05/18/drake-hpc/), and `drake` no longer supports it for remote workers.

The most successful solutions so far have been

1. Totally transient workers with the `"future"` and `"Makefile"` backends, and
2. Totally persistent workers with the `"mclapply"`, `"parLapply"`, and `"future_lapply"` backends. Here, the workers cannot and do not block the master process that sends them instructions. I have found this asynchronicity to be essential for traversing the dependency graph in parallel as efficiently as possible. Whether through `future` or `clustermq` itself, I hope we can make it happen.

**Comment by wlandau** [comment]
I did try to use `clustermq` workers just like the `lapply`-like persistent worker backends, but I found it defeated the purpose. Currently, `drake` uses the file system to communicate with its persistent workers, and those workers use the file system to cache the targets they build. While this works (I took special precautions to avoid race conditions due to NFS file latency) the overhead is large. That's why I hope instructions and targets can be transmitted with ZeroMQ and the file system stuff can be totally left to the master process.

**Comment by mschubert** [comment]
I'm beginning to see the issues you are facing. It should be possible to use `clustermq` workers for individual function calls, while not using the main event loop.

This would be to use something like:

```r
w = workers(2) # submit 2 worker jobs for testing
on.exit(w$cleanup())

msg = w$receive_data()
if (!is.null(msg$result))
    # handle result ...

# if there is work
# this still needs some adjustments on cmq for this to work
w$set_common_data(data=list(fun, const, export, seed)) # set function to call, constant args
w$send_common_data()
w$send_job_data(msg=list(id="do_chunk", df)) # df is data.frame with row=id and fun args

# if not
w$disconnect_worker(msg)
```

Processing the calls this way would be asynchronous, with the `drake` parallel backend responsible for the event loop.

Is that something you could work with?

**Comment by wlandau** [comment]
Thanks for meeting me half way! I think I get the general idea, and I want to understand more about the details.

From what I understand, the master process will loop over something like the above code to receive worker data and send jobs to workers.

After `msg <- w$receive_data()`, what are the contents of `msg`? Is it the output from the worker that finished soonest, results from all the workers, or something else?

In `w$send_job_data()`, would I be sending a single job to the workers collectively? Will `clustermq`'s load balancing decide which worker gets assigned?

**Comment by mschubert** [comment]
I've written about the specification [here](https://github.com/mschubert/clustermq/wiki/Message-specification), hope that helps. Please not the proposed drake-specific extensions `DO_CALL`, which is `DO_SETUP` and `DO_CHUNK` in one command (otherwise there's no guarantee you're talking to the same worker).

> From what I understand, the master process will loop over something like the above code to receive worker data and send jobs to workers.

Yes

> After `msg <- w$receive_data()`, what are the contents of `msg`? Is it the output from the worker that finished soonest, results from all the workers, or something else?

It will be the result of a single chunk (in your case, likely a function call or expression on a single set of arguments).

The fields of `msg` will be: `id`, `result`, `warnings`, and `errors` (as documented above)

> In `w$send_job_data()`, would I be sending a single job to the workers collectively? Will `clustermq`'s load balancing decide which worker gets assigned?

This would be a single call to a single worker. You'd loop through the following way:

```r
switch(msg$id,
    "WORKER_UP" = {
        if (there is calls left that you can do right now)
            w$do_call(...)
        if (there is a result in msg$id)
            # handle result here
    }
    "WORKER_ERROR" = {
        # something went wrong, clean up and print error
    }
)
```

Calls will be load-balanced, but you need to check the result ID to match the result to a call you sent. Otherwise, you'll have no idea which worker sent which result.

**Comment by wlandau** [comment]
Thanks for elaborating. I will experiment with this and get back to you.

**Comment by wlandau** [comment]
Michael, thank you for your patience. I know I was slow getting back to you, and I wanted to make sure I had a good quality block of time to devote to this.

> I've written about the specification here, hope that helps. Please note the proposed drake-specific extensions DO_CALL, which is DO_SETUP and DO_CHUNK in one command (otherwise there's no guarantee you're talking to the same worker).

The spec helped a lot, and I do see the necessity of `DO_CALL`. Thank you.

Regarding your follow-up code from https://github.com/mschubert/clustermq/issues/86#issuecomment-402719971: do you plan to implement `w$do_call(...)` as part of the `DO_CALL` message type? Maybe short for `w$send_job_data(msg = list(id = "DO_CALL")`?

I studied your code from https://github.com/mschubert/clustermq/issues/86#issuecomment-402512794, and the results of my experimentation are below, along with some embedded questions and comments. I am optimistic about interacting with `clustermq` this way, and I have decided to seriously pursue it for `drake::make(parallelism = "clustermq")`. I realize we still need `DO_CALL`, and I would also appreciate your help on 

```r
library(clustermq)
#> * Option 'clustermq.scheduler' not set, defaulting to SGE
#> --- see: https://github.com/mschubert/clustermq/wiki#setting-up-the-scheduler
w <- workers(2)
#> Submitting 2 worker jobs (ID: 7628) ...
on.exit(w$finalize()) # See #91.
msg <- w$receive_data()
msg
#> $id
#> [1] "WORKER_UP"
#>
#> $pkgver
#> [1] 0.8.4.1
if (is.null(msg)){
  # Under what circumstances is the message NULL?
  # What should I check if it is?
}
# From https://github.com/mschubert/clustermq/blob/3c7fda0a7c6ddcc7f9b7ac557a6a2abfcc917e16/R/Q_rows.r#L24-L25
data <- list(
  fun = function(x){
    x * 2
  },
  const = list(),
  export = list(),
  rettype = "list",
  common_seed = 123
)
# From https://github.com/mschubert/clustermq/blob/3c7fda0a7c6ddcc7f9b7ac557a6a2abfcc917e16/R/Q_rows.r#L41
# Do we need `id = "DO_SETUP"` somewhere?
do.call(w$set_common_data, data)
w$send_common_data()
w$send_job_data(
  msg = list(
    id = "DO_CHUNK",       # Maybe eventually "DO_CALL"?
    df = data.frame(x = 1) # drake will deploy targets 1 at a time.
  )
)
#> Operation cannot be accomplished in current state
msg <- w$receive_data()
msg
#> $id
#> [1] "WORKER_ERROR"
#>
#> $token
#> [1] "agasx"
w$disconnect_worker(msg)
```

It seems I do not know how to set the common data properly. In a previous attempt, I tried to send `data` more like you did in https://github.com/mschubert/clustermq/issues/86#issuecomment-402512794, but I get a worker error there too.

```r
w$set_common_data(data)
w$send_common_data()
w$send_job_data(
  msg = list(
    id = "DO_CHUNK",       # Maybe eventually "DO_CALL"?
    df = data.frame(x = 1) # drake will deploy targets 1 at a time.
  )
)
#> Operation cannot be accomplished in current state
msg <- w$receive_data()
msg
#> $id
#> [1] "WORKER_ERROR"
#>
#> $msg
#> [1] "wrong field names for DO_SETUP: data"
```

Like you said, maybe `clustermq` needs more development before this will work. How far can I get with the current development version?

**Comment by mschubert** [comment]
@wlandau Can you try the following using [`develop`](https://github.com/mschubert/clustermq/tree/develop)?

```r
w = workers(1)

# worker connecting to master with id=WORKER_UP
w$receive_data()

# you should be able to send arbitrary expressions to evaluate on a worker
w$send_call(x*2, list(x=5))

# receive the result from any worker with id=WORKER_READY
w$receive_data()

w$cleanup()
```

**Comment by wlandau** [comment]
It worked exactly as I guessed.

```r
> options(clustermq.scheduler = "sge", clustermq.template = "template.tmpl")
> library(clustermq)
> w = workers(1)
Submitting 1 worker jobs (ID: 7225) ...
> # worker connecting to master with id=WORKER_UP
> w$receive_data()
$id
[1] "WORKER_UP"

$pkgver
[1] ‘0.8.4.99’

> # you should be able to send arbitrary expressions to evaluate on a worker
> w$send_call(x*2, list(x=5))
> # receive the result from any worker with id=WORKER_READY
> w$receive_data()
$id
[1] "WORKER_READY"

$token
[1] NA

$expr
x * 2

$result
[1] 10

> w$cleanup()
```

And with a low enough `timeout`, it looks like I can keep calling `w$receive_data()` until a worker is ready to send a result. Multiple calls are giving me a little trouble at the moment, though (maybe I am getting a little too far ahead here).

```r
> options(clustermq.scheduler = "sge", clustermq.template = "template.tmpl")
> library(clustermq)
> 
> w = workers(2)
Submitting 2 worker jobs (ID: 7321) ...
> 
> # worker connecting to master with id=WORKER_UP
> w$receive_data()
$id
[1] "WORKER_UP"

$pkgver
[1] ‘0.8.4.99’

> w$receive_data(timeout = 1e-9)
NULL
> 
> # you should be able to send arbitrary expressions to evaluate on a worker
> w$send_call(x*2, list(x=5))
> w$send_call(x*2, list(x=5))
Operation cannot be accomplished in current state
> 
> # receive the result from any worker with id=WORKER_READY
> w$receive_data()
$id
[1] "WORKER_READY"

$token
[1] NA

$expr
x * 2

$result
[1] 10

> 
> w$cleanup()
Master: [2.3s 1.5% CPU]; Worker: [avg NA% CPU, max NA Mb]
```

**Comment by mschubert** [comment]
For your 2nd example you still need an event loop ;-)

```r
while(work remaining) {
    msg = w$receive_data() # wait until a worker is ready
    result = w$send_call(next work) # send it work if it is
    # ...handle your result...
}
```

The key here is that receive and send are called alternately, so that ZeroMQ can distribute the work in the background.

**Comment by wlandau** [comment]
I see. This is beginning to take shape.

Now that we're getting into more of the details, I recall that the next worker needs to have the right data in memory before its next `send_call()`. Each target has its own set of dependencies, and some of the previous target's dependencies can/should be safely removed. For `drake`'s other parallel backends, I handle this with [`prune_envir()`](https://github.com/ropensci/drake/blob/16700f77700729511db103610f55c23978d7fef7/R/envir.R#L28), but and I think something similar would help reduce the data we send and conserve memory.

Also, `drake` keeps track of a large configuration list internally ([`drake_config()`](https://ropensci.github.io/drake/reference/drake_config.html)), and if possible, I would like to send it to the workers just once at the beginning.

```r
w$send_job_data(config)
while(work remains or worker still running) {
    msg = w$receive_data() # wait until a worker is ready
    if (identical(msg$id, "WORKER_READY")) {
        store_result(msg)
        target <- choose_next_target(msg)
        load_these <- new_dependency_objects(target) # The next worker needs these.
        delete_these <- old_dependency_names(target) # Names of targets that can be safely removed.
        if (!is.null(target)){
            w$
            w$send_call(build_target, list(target = target), const = list(delete_these = delete_these))
        } else {
            w$send_shutdown_worker()
        }
    } else if (identical(msg$id, "WORKER_DONE") {
        w$disconnect_worker()
    } else if (identical(msg$id, "WORKER_ERROR") {
        stop("worker error")
    }
}
```

**Comment by mschubert** [comment]
Looks good!

Some minor pointers:

If you want to send the config separately in the beginning, use `w$set_common_data` with your `config` object in `export` and dummy fields for `fun`, `const`, `rettype`, and `common_seed`.<sup>1</sup> In your loop, you'd check for `WORKER_UP` (respond with `w$send_job_data`), and `WORKER_READY` (do `w$send_call`) as you do now.

`w$send_call` takes an expression (`expr`) and all dependencies in a list (`env`); There is no `const` argument, because for a single call it doesn't make sense to distinguish between constant and iterated.<sup>2</sup>

<hr>

<sup>1</sup> I know requiring dummy fields is not ideal, and this is still subject to change
<sup>2</sup> I'm thinking about generalizing `token` to refer to compute environments, so that workers can perform operations on multiple sets of available data (that were previously exported to that worker); this way you could also track what objects they have in memory

**Comment by wlandau** [comment]
I am not sure I understand your comments about `WORKER_UP` and `w$send_job_data()`, especially since there is no `DO_CHUNK` message in [section of the message specification on evaluating custom expressions](https://github.com/mschubert/clustermq/wiki/Message-specification#evaluating-custom-expressions). What is the role of `w$send_job_data()` here? If `w$send_call()` already has an `env` argument, what additional value does `w$send_job_data()` provide?

Also, do we need a `w$send_common_data()` after `w$set_common_data()`? Do we need calls to `receive_data()` after each?

I am having trouble understanding the functions in the QSys class and when to use them. I think it would be helpful to see in the message specification the message types that each method sends and receives.

**Comment by mschubert** [comment]
`w$set_common_data()` serializes the common data and stores it as a zeromq message object on the master
`w$send_common_data()` sends that object to a worker (and hence we're not serializing and copying memory every time we send common data)
`w$send_job_data()` is for processing a chunk of the same function calls, I think in your use case you shouldn't use that at all

Sorry for the confusion, I'll amend the actual function calls in addition to the message IDs.

**Comment by wlandau** [comment]
Thanks, that helps. With no `w$send_job_data()` in this case, is there any need to handle `WORKER_UP` messages?

**Comment by mschubert** [comment]
Yes: either to send common data after a worker is up, or treating it the same way as `WORKER_READY`

See here: https://github.com/mschubert/clustermq/wiki/Message-specification#evaluating-custom-expressions

**Comment by wlandau** [comment]
That example code in the message specification was really helpful, and I have begun implemneting a `clustermq` backend for `drake` in the [`clustermq` branch](https://github.com/ropensci/drake/tree/clustermq). All the relevant code is in [`R/clustermq.R](https://github.com/ropensci/drake/blob/clustermq/R/clustermq.R), and I have been testing it out with:

```r
library(drake)
load_mtcars_example() # loads the "my_plan" tibble and write report.Rmd
clean(destroy = TRUE) # starts from scratch
# options(clustermq.scheduler = "sge", clustermq.template = "template.tmpl")
options(clustermq.scheduler = "multicore")
make(my_plan, parallelism = "clustermq", jobs = 2)
```

Right now, the multicore backend usually hangs, and the SGE backend usually results in "Interrupted system call" errors from [`receive_data()`](https://github.com/ropensci/drake/blob/27590762b89db742e42f415a2ef01a59e39a0092/R/clustermq.R#L31).

**Comment by mschubert** [comment]
I just tried it, the cmq worker looks fine (evaluates the expression and sends results back).

However, I get stuck endlessly in `mc_get_checksum`. Looks to me like you calculate a checksum per object in message, and then check this instead of only the result you're interested in?

```r
exiting from: mc_get_checksum(target = target, config = config)
debug: if (!identical(local_checksum, checksum)) {
    return(FALSE)
}
Browse[4]> checksum
[1] "01234566814cd2a2b3cd7c19c907964b4a4f018a19ec224788d9edb74d7376df
Browse[4]> local_checksum
[1] "0feaa26f252dbfdd 0feaa26f252dbfdd 4fbaa4449699f7d5 8271c9325d9ef093 01234566814cd2a2b3cd7c19c907964b4a4f018a19ec224788d9edb74d7376df"
Browse[3]> where
where 1: mc_is_good_checksum(target, checksum, config)
where 2: mc_wait_checksum(target = build$target, checksum = build$checksum, config = config)
where 3: cmq_conclude_job(msg = msg, config = config)
where 4: cmq_master(config)
```

**Comment by wlandau** [comment]
Thanks for trying out what I did not get a chance to do Friday afternoon. I only had the morning to implement a first attempt.

I can fix the checksum business. `drake` uses checksums because some targets generate output files that become dependencies of other targets. `clustermq` is so fast that in some cases, if we don't wait for the checksum, some targets may start before their up-to-date dependencies are available.

But this checksum should only apply to reproducibly-tracked output files. `make(parallelism = "clustermq_staged")` already does the right thing, so it should be straightforward to repair.

Thank you for all your help so far!

By the way, what are your plans for development and timing with respect to the next CRAN release?

**Comment by wlandau** [comment]
Update: the new backend has progressed enough for a PR: https://github.com/ropensci/drake/pull/501. The only thing left is an unpredictable yet frequent error:

```r
Error in rzmq::poll.socket(list(private$socket), list("read"), timeout = timeout) :
  Interrupted system call
```

Any ideas about what might be happening?

**Comment by wlandau** [comment]
`txtq` is not actually used for the work that `drake` does with `clustermq`. It is required for `drake`'s `mclapply` backend, which is used by default to process the imports. `make(jobs = c(imports = 1, targets = 2)` and `make(parallelism = c(imports = "mclapply_staged", targets = "clustermq")` do not require `txtq`.

https://github.com/ropensci/rzmq/pull/37 does provide context, thanks.

I am not sure what could be sending the interruption, but since the error message mentions `private$socket`, it seems like the interrupt happens around the [call to `receive_data()`](https://github.com/ropensci/drake/blob/26e871b7b1062c54516aba89ad23321b4f9e9914/R/clustermq.R#L31). I would like to be able to just try `receive_data()` again if there is an interruption, but `try()`/`tryCatch()` statements are ineffective, presumably because of the asynchronous nature of interrupting signals.

**Comment by mschubert** [comment]
Yes, it's no surprise that you catch the interrupts with `receive_data()` because `poll.socket()` is what reacts to them (and you're spending most of your time there).

Using `strace`, I get:

```
--- SIGCHLD {si_signo=SIGCHLD, si_code=CLD_KILLED, si_pid=160983, si_uid=10282396, si_status=SIGTERM, si_utime=1, si_stime=1} ---

Are you running any multicore operations in the background?

**Comment by wlandau** [comment]
Regarding `WORKER_DONE`, I realize that part is my fault. That part of the loop is never reached, and I have been relying on `w$cleanup()`. In the pseudocode from https://github.com/mschubert/clustermq/issues/86#issuecomment-411873939, the `while()` loop checks if a worker is still running. What is the correct way to do that with a `QSys` object?

**Comment by wlandau** [comment]
Never mind: I saw you use `w$workers_running` internally, and it seems to work for `drake` as well. As of https://github.com/ropensci/drake/commit/026cfb32bd940e13037634063bc94d43dda83737, the "Interrupted system call" errors seem to have disappeared completely when I submit to SGE. The multicore backend still shows them when I try am using at least 4 workers at a time, but only when `drake` starts shutting down workers.

**Comment by wlandau** [comment]
https://github.com/mschubert/clustermq/commit/bb4477e1d695143fd2417db35de3b87eae35a442 seemed to help, but [`drake`'s `clustermq` tests]() give several warnings

```r
test-clustermq.R:16: warning: clustermq parallelism
cannot wait for child 14833 as it does not exist
```

Then, sometimes the task hangs, and other times the infamous "Interrupted system call" error appears.

**Comment by wlandau** [comment]
However, jobs on SGE run great, and real HPC schedulers should be `drake`'s main focus for `clustermq`. `drake` already has several local parallel backends, so its only use for multicore `clustermq` is for unit testing. Is there another backend that could allow `drake` to just go through the motions locally?

**Comment by mschubert** [comment]
Please note that the latest version on `develop` no longer sends `WORKER_UP`, but `WORKER_READY` instead.

This will cause problems for [drake/clustermq.R#L34](https://github.com/ropensci/drake/blob/clustermq/R/clustermq.R#L34); you will need to check `token` of the ready message to be the same as you set in common data.

Sorry for this, but that's the downside of using the bleeding edge version. I might still simplify the shutdown procedure before a release (eliminating the requirement for `WORKER_DONE` checks).

**Comment by wlandau** [comment]
Thanks for letting me know. Nothing to apologize for, I am excited to see `clustermq` develop.

In the new `develop` branch, would the workflow be something like this?

```r
cmq_set_common_data <- function(config){
  export <- list()
  if (identical(config$envir, globalenv())){
    export <- as.list(config$envir, all.names = TRUE) # nocov
  }
  export$config <- config
  config$workers$set_common_data(
    export = export,
    fun = identity,
    const = list(),
    rettype = list(),
    common_seed = config$seed,
    token = "set_common_data_token"
  )
}

cmq_master <- function(config){
  while (cmq_work_remains(config) || config$workers$workers_running > 0){
    msg <- config$workers$receive_data()
    cmq_conclude_build(msg = msg, config = config)
    if (identical(msg$id, "WORKER_READY")) {
      if (identical(msg$token, "set_common_data_token")){
        config$workers$send_common_data()
      } else if (cmq_work_remains(config)){
        cmq_send_target(config)
      } else {
        config$workers$send_shutdown_worker()
      }
    } else if (identical(msg$id, "WORKER_DONE")) {
      config$workers$disconnect_worker(msg)
    } else if (identical(msg$id, "WORKER_ERROR")) {
      stop("clustermq worker error") # nocov
    }
  }
}
```

**Comment by wlandau** [comment]
Also, if/when you remove the requirement to check `WORKER_DONE`, what would the cleanup procedure look like?

**Comment by mschubert** [comment]
The idea (implemented in [`clustermq@wapi`](https://github.com/mschubert/clustermq/tree/wapi)) is to handle default events within `workers$receive_data()` and separate the worker API from the message identifiers:

```r
cmq_master <- function(config){
  on.exit(config$workers$finalize())
  while (not all results are in){
    msg <- config$workers$receive_data()
    cmq_conclude_build(msg = msg, config = config)
    if (!identical(msg$token, "set_common_data_token")){
      config$workers$send_common_data()
    } else if (cmq_work_remains(config)){
      cmq_send_target(config)
    } else {
      config$workers$send_shutdown_worker()
    }
  }
  if (config$workers$cleanup())
    on.exit()
}
```

The downside I can see is that if you enter the loop after all results are finished, you will get a `WORKER_DONE` message ID, which you usually don't have to handle (and hence may confuse API users if they don't set their exit condition properly)

**Comment by wlandau** [comment]
I do like having a worker API that does not require checking message types most of the time. I have begun the adjustment in https://github.com/ropensci/drake/commit/8ee84b0cfc7e133a8e2b550f3fc53a3aa49a7b74. Currently, `drake` hangs indefinitely [here at `receive_data()](https://github.com/ropensci/drake/blob/8ee84b0cfc7e133a8e2b550f3fc53a3aa49a7b74/R/clustermq.R#L33) after the first `send_common_data()`.

```r
# devtools::install_github("ropensci/drake", ref = "clustermq")
library(drake)
load_mtcars_example()
clean(destroy = TRUE)
options(clustermq.scheduler = "multicore")
make(my_plan, parallelism = "clustermq", jobs = 2, verbose = 4)
```

**Comment by mschubert** [comment]
You still seem to be using

```r
while (cmq_work_remains(config) || config$workers$workers_running > 0){
```

and not

```r
while (not all results are in){
```

The new API requires that you exit your main loop when the last result is in, not when the last worker is shut down (this is handled by `cleanup()`)

I've added (https://github.com/mschubert/clustermq/commit/2a4c3bfb60ac49f1dfd94659b157faad21c710c1) the errors:

```r
Error in config$workers$receive_data() :
  Trying to receive data without workers
```

and 

```r
Error in config$workers$receive_data() :
  Trying to receive data after work finished
```

to tell the user when they attempt to do that.

**Comment by wlandau** [comment]
Thanks. In https://github.com/ropensci/drake/commit/0489ccfa7667036c22757ea82962f5009725f8ad, the [main loop now checks if all the results are in](https://github.com/ropensci/drake/blob/0489ccfa7667036c22757ea82962f5009725f8ad/R/clustermq.R#L34).

I just tried  https://github.com/mschubert/clustermq/issues/86#issuecomment-414808645 again using https://github.com/ropensci/drake/commit/0489ccfa7667036c22757ea82962f5009725f8ad and https://github.com/mschubert/clustermq/commit/6af6dff7874a1168fbe8e09dc86b9d8f41b8b1cb, and I see the following.

1. The first `receive_data()` succeeds and returns a `WORKER_READY` message.

```r
[1] "receiveing"
$id
[1] "WORKER_READY"

$pkgver
[1] â0.8.4.99â

$token
[1] "not set"
```

2. The first `set_common_data()` runs without error. 
3. The second `receive_data()` quits with the new "Trying to receive data without workers" error.

The results are the same whether I use 1, 2, or 8 workers.

**Comment by wlandau** [comment]
Wow, I can't believe I missed that! Thanks!

**Comment by wlandau** [comment]
Initial testing is looking so much better! #99 aside, my last concern is that I am seeing some (but not all) of my SGE jobs still running even after cleanup. I submitted more workers than targets, so I wonder if these remaining workers never fully started up to begin with.

Anyway, is the new worker API likely to change significantly going forward? If not, we are close being ready to merge https://github.com/ropensci/drake/pull/501.

**Comment by mschubert** [comment]
Good to see that it works & happy I could contribute a bit!

I still aim to solve #99 and have the `rzmq` patch merged before a CRAN release, so I estimate about 10 days for `0.8.5`.

**Comment by wlandau** [comment]
Fantastic! I will keep an eye out for the next CRAN release.
